libsolidity/analysis/ContractLevelChecker.cpp

```diff
@@ -510,3 +671,146 @@ void ContractLevelChecker::checkBaseABICompatibility(ContractDefinition const& _
+	{
+		FunctionSet baseSet = initFrom<FunctionSet>(contract->definedFunctions());
+
+		for (FunctionDefinition const* function: getBaseFunctions(contract))
+		{
+			if (baseSet.cend() == baseSet.find(function))
+				baseSet.insert(function);
+		}
+
+		if (baseSet.empty())
+			continue;
+
+		// Collect the first of every override-set of functions (they are sorted
+		// by most-inherited function)
+		for (auto it = baseSet.cbegin(); it != baseSet.cend(); it = baseSet.upper_bound(*it))
+		{
+			if (!(*it)->isOverridable())
+				continue;
+
+			collectedFunctions.insert(*it);
+		}
+	}
+
+	// Check the sets of the most-inherited functions
+	for (auto it = collectedFunctions.cbegin(); it != collectedFunctions.cend(); it = collectedFunctions.upper_bound(*it))
+	{
+		auto [begin,end] = collectedFunctions.equal_range(*it);
+
+		// Only one function
+		if (next(begin) == end)
+			continue;
+
+		// Not an overridable function
+		if (!(*it)->isOverridable())
+			continue;
+
+		// Function has been explicitly overridden
+		if (contains_if(
+				contractFuncs,
+				[&] (FunctionDefinition const* _f) {
+					return hasEqualNameAndParameters(*_f, **it);
+				}
+			)
+		)
+			continue;
+
+		set<FunctionDefinition const*> ambiguousFunctions;
+		SecondarySourceLocation ssl;
+
+		for (;begin != end; begin++)
+		{
+			ambiguousFunctions.insert(*begin);
+			ssl.append("Definition here: ", (*begin)->location());
+		}
+
+		// Make sure the functions are not from the same base contract
+		if (ambiguousFunctions.size() == 1)
+			continue;
+
+		m_errorReporter.typeError(
+			_contract.location(),
+			ssl,
+			"Functions of the same name " +
+			(*it)->name() +
+			" and parameter types defined in two or more base contracts must be overridden in the derived contract."
+		);
+	}
+}
+
+vector<ContractDefinition const*> ContractLevelChecker::resolveOverrideList(OverrideSpecifier const& _overrides) const
+{
+	vector<ContractDefinition const*> resolved;
+
+	for (ASTPointer<UserDefinedTypeName> const& override: _overrides.overrides())
+	{
+		Declaration const* decl  = override->annotation().referencedDeclaration;
+		solAssert(decl, "Expected declaration to be resolved.");
+
+		// If it's not a contract it will be caught
+		// in the reference resolver
+		if (ContractDefinition const* contract = dynamic_cast<decltype(contract)>(decl))
+			resolved.emplace_back(contract);
+	}
+
+	return resolved;
+}
+
+ContractLevelChecker::FunctionSet const& ContractLevelChecker::getBaseFunctions(ContractDefinition const* _contract) const
+{
+	auto const& result = m_contractBaseFunctions.find(_contract);
+
+	if (result != m_contractBaseFunctions.cend())
+		return result->second;
+
+	FunctionSet set;
+
+	for (auto const* base: resolveBaseContracts(*_contract))
+	{
+		FunctionSet tmpSet = initFrom<FunctionSet>(base->definedFunctions());
+
+		for (auto const& func: getBaseFunctions(base))
+			if (tmpSet.cend() == tmpSet.find(func))
```

Comment:
Why do you have to search first instead of inserting in any case?

Reply:
It's a multiset, so it would append if it already exists, which we don't want in this case. (We only want the most derived variant of a function/modifier)

Reply:
Then why is it a multiset? Wouldn't a regular set be better in this case? (`set` can still be a multiset).

Also please change the typedef to `FunctionMultiset` or not use a typedef at all.

Reply:
It is a multiset because we iterate over all the base contracts to detect multiple inheritance. 
Basically, we add all functions from direct base contracts, but for each base contracts we only add not-yet defined functions from their base contracts.

Will rename the typedef.

Reply:
Will ``tmpSet`` ever contain two keys comparing equal? ``set`` in the outer scope has to be a multiset, but ``tmpSet`` in the inner scope?

Reply:
Right.. the inner can be a normal one. Missed that.

Reply:
Right, I missed that.